teps/0144-param-enum.md

```diff
@@ -0,0 +1,314 @@
+        image: bash
+        script: |
+          # a script to check if $(params.FORMAT) is either `oci` or `docker`.
+      - name: build 
+        image: quay.io/buildah/stable:v1.23.3
+        script: |
+          # build image using buildah
+        ...
+  ```
+
+- As the [gcs-create-bucket](https://github.com/tektoncd/catalog/tree/main/task/gcs-create-bucket/0.1) `Task` Author, I want to specify the valid values (`STANDARD`, `NEARLINE`, `COLDLINE`, or `ARCHIVE` ) of `param` [`storageClass`](https://github.com/tektoncd/catalog/blob/4754df9c260e8f102616170175cb132cfc3c5544/task/gcs-create-bucket/0.1/gcs-create-bucket.yaml#L31C1-L35C22). I want to fail the validation if the input is NOT valid.
+
+- As a `Task`/`Pipeline` Author in an organization, I want to enhance the Task security by declaring upfront the allowed/approved images digests that can be used in the `Task`.
+
+### Requirements
+- `Task`/`Pipeline` Authors should be able to define a set of allowed constants for each param.
+- Tekton should fail the `TaskRun`/`PipelineRun` if user provided value is not in the predefined constants set **before** running the pod.
+- Tekton should give early and explicit error message if the Task/Pipeline is failed due to invalid `param` input
+
+## Proposal
+We propose adding an ***optional*** field named `enum` to the [`ParamSpec`](https://github.com/tektoncd/pipeline/blob/main/docs/pipeline-api.md#paramspec) in a [`PipelineSpec`](https://github.com/tektoncd/pipeline/blob/main/docs/pipeline-api.md#tekton.dev/v1.PipelineSpec) or [`TaskSpec`](https://github.com/tektoncd/pipeline/blob/main/docs/pipeline-api.md#tekton.dev/v1.TaskSpec). This field must be an array with at least one element, where each element is unique.
+
+> :warning: The new API field is introduced as an `alpha` feature
+
+> :warning: This `enum` field should be specified if and only if the parameter type is `string`, which is similar to the existing [`properties`](https://github.com/tektoncd/pipeline/blob/main/docs/pipeline-api.md#paramspec) field that should be only specified for `object` type.
+
+### Tasks
+The following is a `Task` example with the `task-version` param bounded to a set of constants: `v1.21`, `v1.20` and `v1.19`: 
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: golang-build
+spec:
+  params:
+  - name: task-version
+    description: golang version to use for builds
+    type: string
+    enum: # this field can be specified iff the type is string
+      - "v1.21"
+      - "v1.20"
+      - "v1.19"
+    default: "v1.21"
+  steps:
+  - name: build
+    image: docker.io/library/golang:$(params.task-version)
+    ...
+```
+
+#### TaskRuns with TaskRef or Remote Resolution
+`Task` users can provide a value for the parameter at the execution time. If the value is in the `enum` list specified by the author, the task will be executed normally. The task will fail the validation with reason `InvalidParamValue` otherwise. 
```

Comment:
suggest adding a bit more info to `InvalidParamValue` - may be `ParamViolatesDataValidationRestriction"

Reply:
`ParamViolatesDataValidationRestriction` might be a bit too long, I am thinking to add the details in the `message` field instead of `Reason`?